tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -554,6 +554,93 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_io_pair_close(&io_pair));
     };
 
+    /* Test: s2n_connection_get_session should not return a ticket before it has
+     * been received if TLS 1.3 was negotiated.
+     *
+     * This behavior differs between TLS 1.2 and TLS 1.3, so negotiate TLS 1.3 with
+     * a TLS 1.2 session ticket and assert that the TLS 1.3 connection still
+     * behaves correctly.
+     */
+    {
+        DEFER_CLEANUP(struct s2n_blob ticket = { 0 }, s2n_free);
+
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, tls12_server_config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial handshake to produce TLS1.2 session ticket */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_TRUE(IS_ISSUING_NEW_SESSION_TICKET(server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS12);
+
+        /* Store the TLS1.2 session ticket */
+        size_t tls12_session_ticket_len = s2n_connection_get_session_length(client_conn);
+        EXPECT_SUCCESS(s2n_alloc(&ticket, tls12_session_ticket_len));
+        EXPECT_SUCCESS(s2n_connection_get_session(client_conn, ticket.data, ticket.size));
```

Comment:
Nit: unless you JUST declared `ticket`, I'd generally prefer s2n_realloc over s2n_alloc. They have the same result, but if you call s2n_alloc on a blob that already has memory, you'll get a leak. I know you declared `ticket` at the top of this test and it hasn't already been alloc'd, but the gap between the two events makes me nervous :) And this is just a test anyway.

It's a sharp edge, and we should probably deprecate alloc. I think the concern with just blindly replacing all allocs with reallocs was that code might actually require the alloc behavior of preserving the original memory.